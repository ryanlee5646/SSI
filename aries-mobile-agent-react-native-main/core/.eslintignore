node_modules
__tests__
ios
android

# generated by bob
lib/
bifold-app/
generate-app-package-json.js