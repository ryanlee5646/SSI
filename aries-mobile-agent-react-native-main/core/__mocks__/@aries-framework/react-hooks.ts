import { CredentialRecord, ProofRecord } from '@aries-framework/core'

const useCredentials = jest.fn().mockReturnValue({ credentials: [] } as any)
const useCredentialByState = jest.fn().mockReturnValue([] as CredentialRecord[])
const useProofByState = jest.fn().mockReturnValue([] as ProofRecord[])
const mockCredentialModule = {
  acceptOffer: jest.fn(),
  declineOffer: jest.fn(),
}
const mockProofModule = {
  getRequestedCredentialsForProofRequest: jest.fn(),
  acceptRequest: jest.fn(),
  declineRequest: jest.fn(),
}
const useAgent = () => ({
  agent: {
    credentials: mockCredentialModule,
    proofs: mockProofModule,
  },
})
const useCredentialById = jest.fn()
const useProofById = jest.fn()
const useConnectionById = jest.fn()

export {
  useAgent,
  useConnectionById,
  useCredentials,
  useCredentialById,
  useCredentialByState,
  useProofById,
  useProofByState,
}
