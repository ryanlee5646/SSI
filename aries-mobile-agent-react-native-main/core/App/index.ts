/* eslint-disable import/no-cycle */
import type { OnboardingStyleSheet } from './screens/Onboarding'

import { Agent } from '@aries-framework/core'
import AgentProvider from '@aries-framework/react-hooks'

import indyLedgers from '../configs/ledgers/indy'

import Button, { ButtonType } from './components/buttons/Button'
import CheckBoxRow from './components/inputs/CheckBoxRow'
import ErrorModal from './components/modals/ErrorModal'
import InfoTextBox from './components/texts/InfoTextBox'
import toastConfig from './components/toast/ToastConfig'
import RootStack from './navigators/RootStack'
import OnboardingPages from './screens/OnboardingPages'
import Splash from './screens/Splash'
import Terms from './screens/Terms'

export { LocalStorageKeys } from './constants'
export { initLanguages, initStoredLanguage, translationResources } from './localization'
export { ConfigurationProvider, useConfiguration } from './contexts/configuration'
export { StoreProvider, StoreContext } from './contexts/store'
export { DispatchAction } from './contexts/reducers/store'
export { ThemeProvider, useTheme } from './contexts/theme'
export { ColorPallet } from './theme'
export { theme } from './theme'
export { testIdWithKey } from './utils/testable'
export { Screens } from './types/navigators'
export { createStyles } from './screens/OnboardingPages'

export type { StoreProviderProps } from './contexts/store'
export type { Theme } from './theme'
export type { ConfigurationContext } from './contexts/configuration'
export type { Onboarding as OnboardingState } from './types/state'
export type { GenericFn } from './types/fn'
export type { AuthenticateStackParams } from './types/navigators'
export type { OnboardingStyleSheet }

export {
  indyLedgers,
  Agent,
  AgentProvider,
  Button,
  ButtonType,
  CheckBoxRow,
  InfoTextBox,
  ErrorModal,
  toastConfig,
  RootStack,
  OnboardingPages,
  Splash,
  Terms,
}
