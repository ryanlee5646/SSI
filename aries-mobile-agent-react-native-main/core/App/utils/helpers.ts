import {
  ConnectionRecord,
  CredentialMetadataKeys,
  CredentialRecord,
  ProofRecord,
  RequestedAttribute,
} from '@aries-framework/core'
import { useConnectionById, useCredentialById, useProofById } from '@aries-framework/react-hooks'
import { parseUrl } from 'query-string'

import { Attribute } from '../types/record'

export function parseSchema(schemaId?: string): { name: string; version: string } {
  let name = 'Credential'
  let version = ''
  if (schemaId) {
    const schemaIdRegex = /(.*?):([0-9]):([a-zA-Z .\-_0-9]+):([a-z0-9._-]+)$/
    const schemaIdParts = schemaId.match(schemaIdRegex)
    if (schemaIdParts?.length === 5) {
      name = `${schemaIdParts?.[3].replace(/_|-/g, ' ')}`
        .split(' ')
        .map((schemaIdPart) => schemaIdPart.charAt(0).toUpperCase() + schemaIdPart.substring(1))
        .join(' ')
      version = schemaIdParts?.[4]
    }
  }
  return { name, version }
}

export function credentialSchema(credential: CredentialRecord): string | undefined {
  return credential.metadata.get(CredentialMetadataKeys.IndyCredential)?.schemaId
}

export function parsedSchema(credential: CredentialRecord): { name: string; version: string } {
  return parseSchema(credentialSchema(credential))
}

/**
 * Adapted from https://stackoverflow.com/questions/3426404/create-a-hexadecimal-colour-based-on-a-string-with-javascript
 */
export function hashCode(s: string): number {
  return s.split('').reduce((hash, char) => char.charCodeAt(0) + ((hash << 5) - hash), 0)
}

export function hashToRGBA(i: number) {
  const colour = (i & 0x00ffffff).toString(16).toUpperCase()
  return '#' + '00000'.substring(0, 6 - colour.length) + colour
}

export function credentialRecordFromId(credentialId?: string): CredentialRecord | void {
  if (credentialId) {
    return useCredentialById(credentialId)
  }
}

export function connectionRecordFromId(connectionId?: string): ConnectionRecord | void {
  if (connectionId) {
    return useConnectionById(connectionId)
  }
}

export function proofRecordFromId(proofId?: string): ProofRecord | void {
  if (proofId) {
    return useProofById(proofId)
  }
}

export function getConnectionName(connection: ConnectionRecord | void): string | void {
  if (!connection) {
    return
  }
  return connection?.alias || connection?.invitation?.label
}

export function firstAttributeCredential(attributes: RequestedAttribute[], revoked = true): RequestedAttribute | null {
  if (!attributes.length) {
    return null
  }

  let first = null
  const firstNonRevoked = attributes.filter((attribute) => !attribute.revoked)[0]
  if (firstNonRevoked) {
    first = firstNonRevoked
  } else if (attributes.length && revoked) {
    first = attributes[0]
  }

  if (!first?.credentialInfo) {
    return null
  }

  return first
}

export const isRedirection = (url: string): boolean => {
  const queryParams = parseUrl(url).query
  return !(queryParams['c_i'] || queryParams['d_m'])
}

export const processProofAttributes = (
  proof: ProofRecord,
  attributes: Record<string, RequestedAttribute[]> = {}
): Attribute[] => {
  const processedAttributes = [] as Attribute[]
  const { requestedAttributes: requestedProofAttributes } = proof.requestMessage?.indyProofRequest || {}

  requestedProofAttributes?.forEach(({ name, names }, attributeName) => {
    const firstCredential = firstAttributeCredential(attributes[attributeName] || [])
    const credentialAttributes = names?.length ? names : [name || attributeName]
    credentialAttributes.forEach((attribute) => {
      processedAttributes.push({
        name: attribute,
        value: firstCredential?.credentialInfo?.attributes[attribute] || null,
        revoked: firstCredential?.revoked || false,
        credentialId: firstCredential?.credentialId,
      })
    })
  })
  return processedAttributes
}
