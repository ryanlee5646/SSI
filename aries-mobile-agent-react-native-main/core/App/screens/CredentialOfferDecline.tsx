import { useNavigation } from '@react-navigation/core'
import React, { useEffect, useState } from 'react'
import { useTranslation } from 'react-i18next'
import { Modal, StyleSheet, Text, View } from 'react-native'
import { SafeAreaView } from 'react-native-safe-area-context'

import CredentialDeclined from '../assets/img/credential-declined.svg'
import Button, { ButtonType } from '../components/buttons/Button'
import InfoBox, { InfoBoxType } from '../components/misc/InfoBox'
import { useTheme } from '../contexts/theme'
import { GenericFn } from '../types/fn'
import { Screens, TabStacks } from '../types/navigators'
import { testIdWithKey } from '../utils/testable'

export interface CredentialOfferDeclineProps {
  visible: boolean
  didDeclineOffer: boolean
  onGoBackTouched: GenericFn
  onDeclinedConformationTouched: GenericFn
}

const CredentialOfferDecline: React.FC<CredentialOfferDeclineProps> = ({
  visible,
  didDeclineOffer,
  onGoBackTouched,
  onDeclinedConformationTouched,
}) => {
  const { t } = useTranslation()
  const [modalVisible, setModalVisible] = useState<boolean>(false)
  const navigation = useNavigation()
  const { ColorPallet, TextTheme } = useTheme()
  const imageDisplayOptions = {
    fill: ColorPallet.notification.infoText,
    height: 250,
    width: 250,
  }
  const styles = StyleSheet.create({
    container: {
      flexGrow: 1,
      backgroundColor: ColorPallet.brand.primaryBackground,
    },
    image: {
      marginVertical: 66,
    },
    messageContainer: {
      marginHorizontal: 25,
      marginTop: 25,
      alignItems: 'center',
    },
    messageText: {
      fontWeight: 'normal',
      textAlign: 'center',
      marginTop: 90,
    },
    controlsContainer: {
      marginHorizontal: 25,
      marginTop: 25,
    },
  })

  const onDoneTouched = () => {
    navigation.getParent()?.navigate(TabStacks.HomeStack, { screen: Screens.Home })
  }

  useEffect(() => {
    setModalVisible(visible)
  })

  return (
    <Modal visible={modalVisible} transparent={true} animationType={'none'}>
      <SafeAreaView style={[styles.container]}>
        <View style={[{ marginTop: 25 }]}>
          {!didDeclineOffer && (
            <>
              <InfoBox
                notificationType={InfoBoxType.Warn}
                title={'Are you sure you want to decline this credential?'}
                message={'In order to receive the credential offer again, you will need to reapply with the issuer.'}
              />
              <View style={[styles.controlsContainer]}>
                <Button
                  title={t('CredentialOffer.ConfirmDeclineCredential')}
                  accessibilityLabel={t('CredentialOffer.ConfirmDeclineCredential')}
                  testID={testIdWithKey('ConfirmDeclineCredential')}
                  onPress={onDeclinedConformationTouched}
                  buttonType={ButtonType.Primary}
                />
                <View style={[{ marginTop: 10 }]}>
                  <Button
                    title={t('CredentialOffer.AbortDeclineCredential')}
                    accessibilityLabel={t('CredentialOffer.AbortDeclineCredential')}
                    testID={testIdWithKey('AbortDeclineCredential')}
                    onPress={onGoBackTouched}
                    buttonType={ButtonType.Secondary}
                  />
                </View>
              </View>
            </>
          )}

          {didDeclineOffer && (
            <>
              <View style={[styles.messageContainer]}>
                <Text style={[TextTheme.headingThree, styles.messageText]} testID={testIdWithKey('CredentialDeclined')}>
                  {t('CredentialOffer.CredentialDeclined')}
                </Text>
                <CredentialDeclined style={[styles.image]} {...imageDisplayOptions} />
              </View>

              <View style={[styles.controlsContainer]}>
                <Button
                  title={t('Global.Done')}
                  accessibilityLabel={t('Global.Done')}
                  testID={testIdWithKey('Done')}
                  onPress={onDoneTouched}
                  buttonType={ButtonType.Primary}
                />
              </View>
            </>
          )}
        </View>
      </SafeAreaView>
    </Modal>
  )
}

export default CredentialOfferDecline
