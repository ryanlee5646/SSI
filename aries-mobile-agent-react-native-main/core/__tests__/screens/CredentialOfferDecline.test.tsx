import { CredentialRecord, CredentialState } from '@aries-framework/core'
import { useCredentialById } from '@aries-framework/react-hooks'
import { useNavigation } from '@react-navigation/core'
import { render, waitFor, fireEvent } from '@testing-library/react-native'
import fs from 'fs'
import path from 'path'
import React from 'react'

import CredentialOfferDecline from '../../App/screens/CredentialOfferDecline'
import { testIdWithKey } from '../../App/utils/testable'

jest.mock('react-native/Libraries/Animated/NativeAnimatedHelper')
jest.useFakeTimers('legacy')
jest.spyOn(global, 'setTimeout')

const onGoBackTouched = jest.fn()
const onDeclinedConformationTouched = jest.fn()

// const credentialPath = path.join(__dirname, '../fixtures/degree-credential.json')
// const credential = JSON.parse(fs.readFileSync(credentialPath, 'utf8'))
// const credentialRecord = new CredentialRecord(credential)
// // TODO:(jl) Make a fn to revive JSON dates properly and pass to `parse`
// credentialRecord.createdAt = new Date(credentialRecord.createdAt)

// // @ts-ignore
// useCredentialById.mockReturnValue(credentialRecord)

describe('displays a credential decline screen', () => {
  beforeEach(() => {
    onGoBackTouched.mockReset()
    onDeclinedConformationTouched.mockReset()
  })

  test('renders correctly', () => {
    const tree = render(
      <CredentialOfferDecline
        visible={true}
        didDeclineOffer={false}
        onGoBackTouched={onGoBackTouched}
        onDeclinedConformationTouched={onDeclinedConformationTouched}
      />
    )

    const confirmDeclineButton = tree.getByTestId(testIdWithKey('ConfirmDeclineCredential'))
    const abortDeclineButton = tree.getByTestId(testIdWithKey('AbortDeclineCredential'))

    expect(tree).toMatchSnapshot()
    expect(confirmDeclineButton).not.toBeNull()
    expect(abortDeclineButton).not.toBeNull()
  })

  test('user confirms decline', () => {
    const tree = render(
      <CredentialOfferDecline
        visible={true}
        didDeclineOffer={false}
        onGoBackTouched={onGoBackTouched}
        onDeclinedConformationTouched={onDeclinedConformationTouched}
      />
    )

    const confirmDeclineButton = tree.getByTestId(testIdWithKey('ConfirmDeclineCredential'))

    fireEvent(confirmDeclineButton, 'press')

    expect(onGoBackTouched).toBeCalledTimes(0)
    expect(onDeclinedConformationTouched).toBeCalledTimes(1)
  })

  test('user aborts decline', () => {
    const tree = render(
      <CredentialOfferDecline
        visible={true}
        didDeclineOffer={false}
        onGoBackTouched={onGoBackTouched}
        onDeclinedConformationTouched={onDeclinedConformationTouched}
      />
    )

    const abortDeclineButton = tree.getByTestId(testIdWithKey('AbortDeclineCredential'))

    fireEvent(abortDeclineButton, 'press')

    expect(onGoBackTouched).toBeCalledTimes(1)
    expect(onDeclinedConformationTouched).toBeCalledTimes(0)
  })
})
