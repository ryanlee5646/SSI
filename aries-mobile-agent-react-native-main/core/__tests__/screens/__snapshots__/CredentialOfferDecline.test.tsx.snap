// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`displays a credential decline screen renders correctly 1`] = `
<Modal
  animationType="none"
  hardwareAccelerated={false}
  transparent={true}
  visible={true}
>
  <RNCSafeAreaView
    style={
      Array [
        Object {
          "backgroundColor": "#000000",
          "flexGrow": 1,
        },
      ]
    }
  >
    <View
      style={
        Array [
          Object {
            "marginTop": 25,
          },
        ]
      }
    >
      <View
        style={
          Object {
            "backgroundColor": "#313132",
            "borderColor": "#FCBA19",
            "borderRadius": 5,
            "borderWidth": 1,
            "marginHorizontal": 25,
            "minWidth": 700,
            "padding": 10,
          }
        }
      >
        <View
          style={
            Object {
              "flexDirection": "row",
              "paddingHorizontal": 5,
              "paddingTop": 5,
            }
          }
        >
          <View
            style={
              Array [
                Object {
                  "alignSelf": "center",
                  "marginRight": 10,
                },
              ]
            }
          >
            <Text
              allowFontScaling={false}
              style={
                Array [
                  Object {
                    "color": "#FCBA19",
                    "fontSize": 30,
                  },
                  undefined,
                  Object {
                    "fontFamily": "Material Icons",
                    "fontStyle": "normal",
                    "fontWeight": "normal",
                  },
                  Object {},
                ]
              }
            >
              î€‚
            </Text>
          </View>
          <Text
            style={
              Object {
                "alignSelf": "center",
                "color": "#FFFFFF",
                "flexShrink": 1,
                "fontSize": 18,
                "fontWeight": "bold",
              }
            }
            testID="com.ariesbifold:id/HeaderText"
          >
            Are you sure you want to decline this credential?
          </Text>
        </View>
        <View
          style={
            Object {
              "flexDirection": "column",
              "marginLeft": 40,
              "paddingBottom": 5,
              "paddingHorizontal": 5,
            }
          }
        >
          <Text
            style={
              Object {
                "color": "#FFFFFF",
                "flexShrink": 1,
                "fontSize": 18,
                "fontWeight": "normal",
                "marginVertical": 16,
              }
            }
            testID="com.ariesbifold:id/BodyText"
          >
            In order to receive the credential offer again, you will need to reapply with the issuer.
          </Text>
        </View>
      </View>
      <View
        style={
          Array [
            Object {
              "marginHorizontal": 25,
              "marginTop": 25,
            },
          ]
        }
      >
        <View
          accessibilityLabel="CredentialOffer.ConfirmDeclineCredential"
          accessibilityState={
            Object {
              "disabled": false,
            }
          }
          accessible={true}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "backgroundColor": "#42803E",
              "borderRadius": 4,
              "opacity": 1,
              "padding": 16,
            }
          }
          testID="com.ariesbifold:id/ConfirmDeclineCredential"
        >
          <Text
            style={
              Array [
                Object {
                  "color": "#FFFFFF",
                  "fontSize": 18,
                  "fontWeight": "bold",
                  "textAlign": "center",
                },
                false,
              ]
            }
          >
            CredentialOffer.ConfirmDeclineCredential
          </Text>
        </View>
        <View
          style={
            Array [
              Object {
                "marginTop": 10,
              },
            ]
          }
        >
          <View
            accessibilityLabel="CredentialOffer.AbortDeclineCredential"
            accessibilityState={
              Object {
                "disabled": false,
              }
            }
            accessible={true}
            focusable={true}
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              Object {
                "borderColor": "#42803E",
                "borderRadius": 4,
                "borderWidth": 2,
                "opacity": 1,
                "padding": 16,
              }
            }
            testID="com.ariesbifold:id/AbortDeclineCredential"
          >
            <Text
              style={
                Array [
                  Object {
                    "color": "#42803E",
                    "fontSize": 18,
                    "fontWeight": "bold",
                    "textAlign": "center",
                  },
                  false,
                ]
              }
            >
              CredentialOffer.AbortDeclineCredential
            </Text>
          </View>
        </View>
      </View>
    </View>
  </RNCSafeAreaView>
</Modal>
`;
